api/src/main/java/com/datatorrent/api/DAG.java

```diff
@@ -288,4 +288,23 @@
   {
 
   }
+
+  /**
+   * This class provides functionality to extend existing DAG. The type of extensions allowed are
+   *
+   * <li>Add a new operator</li>
+   * <li>Remove an existing operator</li>
+   * <li>Add an operator to stream</li>
+   * <li>Add an stream between existing operator and new operator</li>
+   */
+  interface DAGChangeSet extends DAG
+  {
+    void removeOperator(String name);
+
+    void removeStream(String name);
+
+    StreamMeta extendStream(String id, Operator.InputPort... ports);
+
+    StreamMeta addStream(String id, String operatorName, String portName, Operator.InputPort... ports);
```

Comment:
While applying the changes the changes are applied on the cloned logical plan first avoid any invalid operations on current plan. Only after the validation is successful the current plan is changed. The port objects are different in the cloned plan hence could not use the reference to port object directly.